### PR TITLE
Change implementation of outStream for Central API Client functions

### LINKED ISSUE #31711: `CentralAPIClient` writing to stdout breaks language server in online mode
**Description:**
When the `CentralAPIClient` is pulling a module, it prints the success message (and progress) to stdout. Since language server and vscode language client communicates over the `stdin/stdout` using json RPC, above behavior cause non-json content to be written into stdout. This breaks the language server protocol and cause LS to not work there onward.

Changing vscoe - LS to communicate over a different stream is a large refactoring + any component inside LS (including project API, compiler, etc) should not be writing to `stdout` when they are invoked via the LS.

Therefore, need a way to change the `outputStream` in `CentralAPIClient` for LS use cases.

**Steps to reproduce:**
See description

**Affected Versions:**
SL Beta 3 SNAPSHOT

**Related Issues (optional):**
#30757

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Add option to disable 'Printing to console' for Central API Client functions.

Inorder to fix the following issue :

> When the `CentralAPIClient` is pulling a module, it prints the success message (and progress) to stdout. Since language server and vscode language client communicates over the stdin/stdout using JSON RPC, above behavior cause non-json content to be written into stdout. This breaks the language server protocol and causes LS to not work there onward.

Fixes #31711

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Comment by Ibaqu** [comment]
Push package and Pull package which makes use of outStream has been modified as such

A boolean `disableOutStream` has been added which will disable all prints to the console.

```
public void pushPackage(Path balaPath, String org, String name, String version, String supportedPlatform,
                            String ballerinaVersion, boolean disableOutStream)
                            
public void pullPackage(String org, String name, String version, Path packagePathInBalaCache,
                            String supportedPlatform, String ballerinaVersion, boolean isBuild,
                            boolean disableOutStream)                            
```

Encased all out.println() with a check to prevent it from being printed. Both functions are overloaded with the defaults as `false` to prevent other features from breaking

File: cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java [hunk 1]
```diff
@@ -274,6 +280,23 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
...
+        byte[] buffer = new byte[1024];
+
+        try {
+            while ((count = inputStream.read(buffer)) > 0) {
+                outputStream.write(buffer, 0, count);
+            }
+        } catch (IOException e) {
+            PrintStream printStream = System.out;
+
+            printStream.println(logFormatter.formatLog(fullPkgName + "pulling the package from central failed"));
```
Comment (hevayo) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:295]: At this level this should be an exception and can be handled at a higher level. Then the exception can be converted to a diagnostic at the Package resolution

File: cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java [hunk 2]
```diff
@@ -274,6 +280,23 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
...
+
+        }
+    }
+
     /**
      * Create nightly build meta file.
      *
```
Comment (hevayo) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:302]: Can we add some tests how can you disable printing when using ProjectAPI ?

**Review by hevayo — CHANGES_REQUESTED** [review]

File: cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java [hunk 3]
```diff
@@ -274,6 +278,17 @@ private static void writeAndHandleProgress(InputStream inputStream, FileOutputSt
         }
     }
 
+    private static void writeAndHandleProgressWithoutLogging(InputStream inputStream, FileOutputStream outputStream)
+            throws IOException {
+        int count;
+        byte[] buffer = new byte[1024];
+
+        while ((count = inputStream.read(buffer)) > 0) {
+            outputStream.write(buffer, 0, count);
```
Comment (azinneera) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:287]: We should not write to the output stream. The idea is that we do the pull operation quietly. If the pull fails due to any interruption, then we should throw an exception and log it as a diagnostic.

File: cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java (hunk 2 above)
Comment (Ibaqu) [reply cli/central-client/src/main/java/org/ballerinalang/central/client/Utils.java:293]: Added tests. Will add tests to distribution as well.

**Comment by Ibaqu** [comment]
Changes : 

- Change `DISABLE_OUTPUT_STREAM` -> `ENABLE_OUTPUT_STREAM` 
- Added `System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "false");` to `CompileTask` disabling output stream for all commands
- Added tests

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java [hunk 4]
```diff
@@ -61,6 +62,8 @@ public void execute(Project project) {
         // Print the source
         this.out.println("\t" + sourceName);
 
+        System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "false");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java:65]: The progress bar should show in the stdout in the case of commands.

Comment (Ibaqu) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java]: Changed to true as discussed

File: cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java [hunk 5]
```diff
@@ -342,8 +347,14 @@ public void pushPackage(Path balaPath, String org, String name, String version,
         }
     }
 
+    /**
+     * Pulling a package.
```
Comment (azinneera) [inline cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java:351]: Shall we refactor the method comment to use proper tense and add the parameters and exceptions thrown?

Comment (Ibaqu) [reply cli/central-client/src/main/java/org/ballerinalang/central/client/CentralAPIClient.java]: Addressed in https://github.com/ballerina-platform/ballerina-lang/commit/a26d8a97e757d634eba17907656536dfe134e2d5

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java [hunk 6]
```diff
@@ -61,6 +62,8 @@ public void execute(Project project) {
         // Print the source
         this.out.println("\t" + sourceName);
 
+        System.setProperty(CentralClientConstants.ENABLE_OUTPUT_STREAM, "true");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/task/CompileTask.java:65]: We need to set this in `PullCommand` and `PushCommand` as well

File: cli/central-client/src/test/java/org/ballerinalang/central/client/TestCentralApiClient.java [hunk 7]
```diff
@@ -526,4 +527,91 @@ private void setBallerinaHome() {
             System.setProperty(ballerinaInstallDirProp, String.valueOf(Paths.get("build")));
         }
     }
+
+    @Test(description = "Test pull package without Logs", enabled = false)
```
Comment (azinneera) [inline cli/central-client/src/test/java/org/ballerinalang/central/client/TestCentralApiClient.java:531]: We can move these tests to distribution without having it disabled here.

Comment (Ibaqu) [reply cli/central-client/src/test/java/org/ballerinalang/central/client/TestCentralApiClient.java]: Removed disabled tests

**Comment by Ibaqu** [comment]
Test : https://github.com/ballerina-platform/ballerina-distribution/pull/2070

**Review by azinneera — APPROVED** [review]

**Review by hevayo — APPROVED** [review]

**Pull request merged**